concent_api/core/payments/report.py

```diff
@@ -0,0 +1,202 @@
+from typing import List
+from typing import NamedTuple
+from typing import Tuple
+
+from django.db.models import QuerySet
+
+from common.helpers import deserialize_message
+from common.helpers import get_current_utc_timestamp
+from common.helpers import parse_datetime_to_timestamp
+from core.exceptions import SCINotSynchronized
+from core.models import Client
+from core.models import PendingResponse
+from core.models import Subtask
+from core.payments import service
+from core.payments.backends.sci_backend import TransactionType
+
+
+class PendingResponses(NamedTuple):
+    subtask_results_settled: QuerySet
+    subtask_results_rejected: QuerySet
+    force_payment_committed: QuerySet
+
+    def is_empty(self) -> bool:
+        is_subtask_results_settled_empty = len(self.subtask_results_settled) == 0
+        is_subtask_results_rejected_empty = len(self.subtask_results_rejected) == 0
+        is_force_payment_committed_empty = len(self.force_payment_committed) == 0
+
+        return is_subtask_results_settled_empty and is_subtask_results_rejected_empty and is_force_payment_committed_empty
+
+
+class Payments(NamedTuple):
+    regular_payments: list
+    settlement_payments: list
+    forced_subtask_payments: list
+    additional_verification_payments_for_requestor: list
+    additional_verification_payments_for_provider: list
+
+    def is_empty(self) -> bool:
+        is_regular_payments_empty = len(self.regular_payments) == 0
+        is_settlement_payments_empty = len(self.settlement_payments) == 0
+        is_forced_subtask_payments_empty = len(self.forced_subtask_payments) == 0
+        is_additional_verification_payments_for_requestor_empty = len(self.additional_verification_payments_for_requestor) == 0
+        is_additional_verification_payments_for_provider_empty = len(self.additional_verification_payments_for_provider) == 0
+
+        return is_regular_payments_empty and is_settlement_payments_empty and is_forced_subtask_payments_empty and is_additional_verification_payments_for_requestor_empty and is_additional_verification_payments_for_provider_empty
+
+
+class PairReport(NamedTuple):
+    requestor: Client
+    provider: Client
+    requestor_ethereum_addresses: list
+    provider_ethereum_addresses: list
+    pending_responses: PendingResponses
+    payments: Payments
+
+
+class ClientEthereumAddresses(NamedTuple):
+    provider: str
+    requestor: str
+
+
+class Report(NamedTuple):
+    pair_reports: list
+
+
+def create_report() -> Report:
+    """ Create payments report for Concent. """
+    pair_reports = []
+
+    for subtask in get_subtasks_with_distinct_clients_pairs():  # pylint: disable=not-an-iterable
+        (subtasks_ids, clients_ethereum_addresses) = get_subtask_ids_and_ethereum_addresses_for_clients_pair(
```

Comment:
I think the code is too compact in this function, and it would be good to separate logical blocks by empty lines.

Reply:
Done